File: heaps/0502_hard/ipo.py
import heapq


def maximize_capital(k: int, wealth: int, profits: list[int], capitals: list[int]) -> int:  # LeetCode 502.
    """Select at most k distinct projects to maximize wealth."""
    capitals_heap: list[tuple[int, int]] = []  # Min heap. Format: (capital, idx).
    for idx, iter_capital in enumerate(capitals):
        heapq.heappush(capitals_heap, (iter_capital, idx))

    profits_heap = []  # Max heap.
    while k > 0:
        while capitals_heap and wealth >= capitals_heap[0][0]:
            idx = heapq.heappop(capitals_heap)[1]
            heapq.heappush(profits_heap, -profits[idx])  # Negate profit to fit into max heap.

        if not profits_heap:
            break
        wealth -= heapq.heappop(profits_heap)  # Negate profit back to original value.
        k -= 1

    return wealth

File: heaps/0502_hard/ipo.cpp
#include <vector>
#include <queue>
using namespace std;

int maximizeCapital(int k, int wealth, vector<int> &profits, vector<int> &capital)
{ // LeetCode 502.
    // Min heap. Format: {capital, profit idx}.
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> capitalsHeap;

    for (int idx = 0; idx < capital.size(); idx++)
        capitalsHeap.push({capital[idx], idx});

    priority_queue<int> profitsHeap; // Max heap.

    while (k > 0)
    {
        while (!capitalsHeap.empty() && wealth >= capitalsHeap.top().first)
        {
            // Grab all the affordable projects' profits into max heap.
            auto [capital, profit_idx] = capitalsHeap.top();
            profitsHeap.push(profits[profit_idx]);
            capitalsHeap.pop();
        }

        if (profitsHeap.empty()) // No more affordable projects.
            break;

        else
        {
            wealth += profitsHeap.top();
            k -= 1;
            profitsHeap.pop();
        }
    }

    return wealth;
}
